Tensile/KernelWriterAssembly.py

```diff
@@ -3021,6 +3020,148 @@ def getKernArg(self, parmName, writeSgpr=1):
+      kStr += self.getKernArg("MagicNumberSize%s"%magicName)
+      kStr += self.getKernArg("MagicShiftSize%s"%magicName)
+
+    for idxChar in kernel["PackedC0IdxChars"][:-1]:
+      kStr += self.getKernArg("MagicNumberSize%s"%idxChar)
+      kStr += self.getKernArg("MagicShiftSize%s"%idxChar)
+    for idxChar in kernel["PackedC1IdxChars"][:-1]:
+      kStr += self.getKernArg("MagicNumberSize%s"%idxChar)
+      kStr += self.getKernArg("MagicShiftSize%s"%idxChar)
+
+    for idx in kernel["ProblemType"]["IndicesSummation"]:
+      for tc in ('A','B'):
+        for zp in kernel["ProblemType"]["ZeroPad%s"%tc]:
+          (freeDim, sumDim, padStart, padEnd) = zp
+          if sumDim == idx:
+            freeDimChar = globalParameters["IndexChars"][freeDim]
+            sumDimChar  = globalParameters["IndexChars"][sumDim]
+            kStr += self.getKernArg("PadStart%s%s%s"%(tc, freeDimChar, sumDimChar))
+            kStr += self.getKernArg("PadEnd%s%s%s"%(tc, freeDimChar, sumDimChar))
+
+    kStr += self.getKernArg("OrigStaggerUIter", self.staggerU)
+
+    kStr += self.getKernArg("NumWorkGroups0")
+    kStr += self.getKernArg("NumWorkGroups1")
+    kStr += self.getKernArg("MagicNumberProblemNumGroupTiles0", kernel["PersistentKernel"])
+    kStr += self.getKernArg("GridNumWorkGroups0", kernel["PersistentKernel"])
+    kStr += self.getKernArg("NumFullBlocks")
+    kStr += self.getKernArg("WgmRemainder1")
+    kStr += self.getKernArg("MagicNumberWgmRemainder1")
+
+    return kStr
+
+
+  ##############################################################################
+  # code phrase for load batched address from array of buffer pointer
+  ##############################################################################
+  def loadBatchedAddress(self, kernel, Batch):
+    kStr = self.endLine
+
+    for idx in kernel["ProblemType"]["IndicesBatch"]:
+      if not isPackedIndex(kernel,idx):
+        tmpSgpr = self.getTmpSgpr(1).idx()
+        kStr += inst("s_mul_i32", sgpr(tmpSgpr), sgpr(Batch), 0x8, "offset of global buffer address")
+        if not kernel["_GlobalAccumulation"]:
+          kStr += inst("s_load_dwordx2", sgpr("AddressD", 2), sgpr("AddressD",2), sgpr(tmpSgpr), "load global buffer D address")
+          kStr += inst("s_load_dwordx2", sgpr("AddressC", 2), sgpr("AddressC",2), sgpr(tmpSgpr), "load global buffer C address")
+        kStr += inst("s_load_dwordx2", sgpr("AddressA", 2), sgpr("AddressA",2), sgpr(tmpSgpr), "load global buffer A address")
+        kStr += inst("s_load_dwordx2", sgpr("AddressB", 2), sgpr("AddressB",2), sgpr(tmpSgpr), "load global buffer B address")
+        kStr += inst("s_waitcnt", "lgkmcnt(0)", "wait global buffer adress ready")
+
+    return kStr
+
+
+  ##############################################################################
+  # code phrase for add buffer offset to address
+  ##############################################################################
+  def addOffsetToAddress(self, kernel):
```

Comment:
In general batch GEMM, input buffers are pointer of buffer array.
So we have to add offset in kernel due to CPU can't access gpu buffer.

There are two place need to add buffer offset, so I use a function to avoid replicate code.
Should I just replicate same code in tow place??

Reply:
No, the function is a good idea, it should just either take some argument that specifies *which* offset and *which* address, or it should have those in the function name.

Reply:
I will put implements back to caller due to some persistent kernel modification